mm2src/kdf_walletconnect/src/session/mod.rs

```diff
@@ -0,0 +1,377 @@
+use wc_common::SymKey;
+
+pub(crate) const FIVE_MINUTES: u64 = 5 * 60;
+pub(crate) const THIRTY_DAYS: u64 = 30 * 24 * 60 * 60;
+
+pub(crate) type WcRequestResponseResult = MmResult<(Value, IrnMetadata), WalletConnectError>;
+
+/// In the WalletConnect protocol, a session involves two parties: a controller
+/// (typically a wallet) and a proposer (typically a dApp). This enum is used
+/// to distinguish between these two roles.
+#[derive(Debug, Display, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum SessionType {
+    /// Represents the controlling party in a session, typically a wallet.
+    Controller,
+    /// Represents the proposing party in a session, typically a dApp.
+    Proposer,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+pub struct SessionRpcInfo {
+    pub topic: Topic,
+    pub metadata: Metadata,
+    pub peer_pubkey: String,
+    pub pairing_topic: Topic,
+    pub namespaces: BTreeMap<String, Namespace>,
+    pub subscription_id: SubscriptionId,
+    pub properties: Option<SessionProperties>,
+    pub expiry: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct KeyInfo {
+    pub chain_id: String,
+    pub name: String,
+    pub algo: String,
+    pub pub_key: String,
+    pub address: String,
+    pub bech32_address: String,
+    pub ethereum_hex_address: String,
+    pub is_nano_ledger: bool,
+    pub is_keystone: bool,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct SessionProperties {
+    #[serde(default, deserialize_with = "deserialize_keys_from_string")]
+    pub keys: Option<Vec<KeyInfo>>,
+}
+
+fn deserialize_keys_from_string<'de, D>(deserializer: D) -> Result<Option<Vec<KeyInfo>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    #[derive(Deserialize)]
+    #[serde(untagged)]
+    enum KeysField {
+        String(String),
+        Vec(Vec<KeyInfo>),
+        None,
+    }
+
+    match KeysField::deserialize(deserializer)? {
+        KeysField::String(key_string) => serde_json::from_str(&key_string)
+            .map(Some)
+            .map_err(serde::de::Error::custom),
+        KeysField::Vec(keys) => Ok(Some(keys)),
+        KeysField::None => Ok(None),
+    }
+}
+
+/// This struct is typically used in the core session management logic of a WalletConnect
+/// implementation. It's used to store, retrieve, and update session information throughout
+/// the lifecycle of a WalletConnect connection.
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+pub struct Session {
+    /// Session topic
+    pub topic: Topic,
+    /// Pairing subscription id.
+    pub subscription_id: SubscriptionId,
+    /// Session symmetric key
+    pub session_key: SessionKey,
+    /// Information about the controlling party (typically a wallet).
+    pub controller: Controller,
+    /// Information about the proposing party (typically a dApp).
+    pub proposer: Proposer,
```

Comment:
this doesn't seem to be ever used?
also why we even have info about both proposer & controller at the same time (so 2 fields)? why not have an enum store info for the opposite party.

Reply:
It's good to save this information, should be use to the GUI